peps/pep-0740.rst

```diff
@@ -0,0 +1,288 @@
+released in 2018.
+
+Additionally, wheel signatures do not satisfy any of
+the above considerations (due to the "attached" nature of the signatures,
+non-verifiability on the index itself, and support for wheels only).
+
+Specification
+=============
+
+Upload endpoint changes
+-----------------------
+
+The current upload API is not standardized. However, we propose the following
+changes to it:
+
+* In addition to the current top-level ``content`` and ``gpg_signature`` fields,
+  the index **SHALL** accept ``attestations`` as an additional multipart form
+  field.
+* The new ``attestations`` field **SHALL** be a JSON object.
+* The JSON object **SHALL** have one or more keys, each identifying an
+  attestation format known to the index. If any key does not identify an
+  attestation format known to the index, the index **MUST** reject the upload.
+* The value associated with each well-known key **SHALL** be a JSON object.
+* Each attestation value **MUST** be verifiable by the index. If the index fails
+  to verify any attestation in ``attestations``, it **MUST** reject the upload.
+
+In addition to the above, the index **SHALL** enforce a consistency
+policy for release attestations via the following:
+
+* If the first file under a new release is supplied with ``attestations``,
+  then all subsequently uploaded files under the same release **MUST** also
+  have ``attestations``. Conversely, if the first file under a new release
+  does not have any ``attestations``, then all subsequent uploads under the
+  same release **MUST NOT** have ``attestations``.
+* All files under the same release **MUST** have the same set of well-known
+  attestation format keys.
+
+The index **MUST** reject any file upload that does not satisfy these
+consistency properties.
+
+Index changes
+-------------
+
+Simple Index
+^^^^^^^^^^^^
+
+* When an uploaded file has an attestation, the index **MAY** include a
+  ``data-attestations`` attribute on its file link, with a value of either
+  ``true`` or ``false``.
+* When ``data-attestations`` is ``true``, the index **MUST** serve the uploaded
+  file's attestations at the same URL, but with ``.attestations`` appended to
+  it. For example, if ``HolyGrail-1.0.tar.gz`` existed and had associated
+  attestations, those attestations would be located at
+  ``HolyGrail-1.0.tar.gz.attestations``.
+* The attestations served by the index **SHALL** be in JSON format, with the
+  same well-known keys as accepted by the upload endpoint.
+
+.. TODO: Represent the trusted publisher metadata here, somehow.
+
+JSON-based Simple API
+^^^^^^^^^^^^^^^^^^^^^
+
+* When an uploaded file has an attestation, the index **MAY** include an
+  ``attestations`` key in the ``file`` dictionary for that file.
+* The ``attestations`` key, when present, **MUST** be a JSON object with
+  the same well-known keys and values as accepted by the upload endpoint.
+
+In addition to the above, when an uploaded file was uploaded via
```

Comment:
Hmm, great question. My first thought is "yes" since the TP metadata is conceptually not a part of any attestation, but I see your point about people leaning on it incorrectly.

Maybe we could have something like this (YAML as psuedocode, would actually be JSON):

```yaml
provenance:
  publisher: { ... }
  attestations: []
```

...that way both the TP metadata and all attestations are present under the same singular key in the JSON API. WDYT?

Reply:
Thought about this some more, and I think this is the way to go. I'll make the appropriate changes today.

Reply:
Done in 25723eec: I've created a "provenance objects" section that explains a JSON object containing both one or more attestations *and* a trusted publisher configuration blob.